File: 222.count-complete-tree-nodes.java
/*
 * @lc app=leetcode id=222 lang=java
 *
 * [222] Count Complete Tree Nodes
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int countNodes(TreeNode root) {
        int h = height(root);
        if (h == -1) return 0;
        return height(root.right) == h-1 ? 
            (1 << h) + countNodes(root.right) :
            (1 << (h-1)) + countNodes(root.left);
    }

    private int height(TreeNode root) {
        if (root == null) return -1;
        return 1 + height(root.left);
    }
}

File: go/222.count-complete-tree-nodes.go
/*
 * @lc app=leetcode id=222 lang=golang
 *
 * [222] Count Complete Tree Nodes
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func countNodes(root *TreeNode) int {
	left, right := root, root
	lh, rh := 0, 0
	for left != nil {
		left = left.Left
		lh++
	}
	for right != nil {
		right = right.Right
		rh++
	}
	if lh == rh {
		return (1 << lh) - 1
	}
	return 1 + countNodes(root.Left) + countNodes(root.Right)
}
// @lc code=end

